File: server-go/internal/codexlb/snapshot.go
// Package codexlb asks a local codex-lb instance for aggregate upstream quota
// usage and normalizes it into Token Terrier's Codex snapshot shape.
package codexlb

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"math"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/codemoo/token-terrier/server-go/internal/wire"
)

const (
	disableEnv = "TOKEN_USAGE_DISABLE_CODEX_LB"
	defaultURL = "http://127.0.0.1:2455"
)

// Snapshotter reads codex-lb's self-service usage endpoint.
type Snapshotter struct {
	BaseURL string
	APIKey  string
	Client  *http.Client

	producer wire.ProducerInfo
	logger   *slog.Logger
}

// NewSnapshotter builds a Snapshotter. The URL defaults to the local codex-lb
// server and can be overridden with TOKEN_USAGE_CODEX_LB_URL or
// CODEX_LB_BASE_URL. API keys are read from TOKEN_USAGE_CODEX_LB_API_KEY first,
// then CODEX_LB_API_KEY.
func NewSnapshotter(producer wire.ProducerInfo, logger *slog.Logger) *Snapshotter {
	if logger == nil {
		logger = slog.Default()
	}
	baseURL := firstNonEmpty(
		os.Getenv("TOKEN_USAGE_CODEX_LB_URL"),
		os.Getenv("CODEX_LB_BASE_URL"),
		defaultURL,
	)
	apiKey := firstNonEmpty(
		os.Getenv("TOKEN_USAGE_CODEX_LB_API_KEY"),
		os.Getenv("CODEX_LB_API_KEY"),
	)
	return &Snapshotter{
		BaseURL:  normalizeBaseURL(baseURL),
		APIKey:   strings.TrimSpace(apiKey),
		Client:   &http.Client{Timeout: 5 * time.Second},
		producer: producer,
		logger:   logger,
	}
}

// Snapshot returns a normalized Codex snapshot from codex-lb if available.
// ok=false means "no codex-lb data, fall back to the normal Codex API path".
func (s *Snapshotter) Snapshot(ctx context.Context, seq int, now time.Time) (snap wire.UsageSnapshot, ok bool) {
	if os.Getenv(disableEnv) == "1" {
		return wire.UsageSnapshot{}, false
	}
	if strings.TrimSpace(s.APIKey) == "" {
		return wire.UsageSnapshot{}, false
	}
	resp, err := s.fetchUsage(ctx)
	if err != nil {
		s.logger.Debug("codex-lb usage fetch failed", "err", err)
		return wire.UsageSnapshot{}, false
	}
	snap, ok = buildSnapshot(resp, seq, s.producer, now)
	if !ok {
		return wire.UsageSnapshot{}, false
	}
	return snap, true
}

func (s *Snapshotter) fetchUsage(ctx context.Context) (usageResponse, error) {
	if strings.TrimSpace(s.BaseURL) == "" {
		return usageResponse{}, errors.New("empty codex-lb URL")
	}
	endpoint := strings.TrimRight(s.BaseURL, "/") + "/v1/usage"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint, nil)
	if err != nil {
		return usageResponse{}, err
	}
	req.Header.Set("Authorization", "Bearer "+s.APIKey)
	req.Header.Set("Accept", "application/json")

	client := s.Client
	if client == nil {
		client = http.DefaultClient
	}
	httpResp, err := client.Do(req)
	if err != nil {
		return usageResponse{}, err
	}
	defer httpResp.Body.Close()

	if httpResp.StatusCode < 200 || httpResp.StatusCode >= 300 {
		return usageResponse{}, fmt.Errorf("codex-lb usage status %d", httpResp.StatusCode)
	}
	var decoded usageResponse
	if err := json.NewDecoder(httpResp.Body).Decode(&decoded); err != nil {
		return usageResponse{}, err
	}
	return decoded, nil
}

type usageResponse struct {
	UpstreamLimits []upstreamLimit `json:"upstream_limits"`
}

type upstreamLimit struct {
	LimitType      string  `json:"limit_type"`
	LimitWindow    string  `json:"limit_window"`
	MaxValue       float64 `json:"max_value"`
	CurrentValue   float64 `json:"current_value"`
	RemainingValue float64 `json:"remaining_value"`
	ModelFilter    *string `json:"model_filter"`
	ResetAt        *string `json:"reset_at"`
	Source         string  `json:"source"`
}

func buildSnapshot(resp usageResponse, seq int, producer wire.ProducerInfo, now time.Time) (wire.UsageSnapshot, bool) {
	if len(resp.UpstreamLimits) == 0 {
		return wire.UsageSnapshot{}, false
	}

	rolling := wire.EmptyRollingWindow()
	weekly := wire.EmptyRollingWindow()
	quotaWindows := make([]wire.QuotaWindow, 0)
	seen := false

	for _, limit := range resp.UpstreamLimits {
		if !isAggregateCreditLimit(limit) || limit.MaxValue <= 0 {
			continue
		}
		usedPct := clamp(limit.CurrentValue/limit.MaxValue, 0, 1)
		resetAt := parseResetAt(limit.ResetAt)
		window := normalizedWindow(limit.LimitWindow)
		switch window {
		case "5h":
			rolling = rollingWindow(usedPct, resetAt, now)
			seen = true
		case "7d":
			weekly = rollingWindow(usedPct, resetAt, now)
			seen = true
		default:
			quotaWindows = append(quotaWindows, quotaWindow(window, usedPct, resetAt))
			seen = true
		}
	}
	if !seen {
		return wire.UsageSnapshot{}, false
	}

	loginMethod := "codex-lb"
	return wire.UsageSnapshot{
		Schema:           1,
		Seq:              seq,
		GeneratedAtUTC:   wire.FormatTime(now),
		ProducerID:       producer.ID,
		ProducerTimeZone: producer.TimeZone,
		Provider:         wire.ProviderCodex,
		BurnState:        "idle",
		Rolling5h:        rolling,
		Weekly:           weekly,
		QuotaWindows:     quotaWindows,
		Credits:          nil,
		Extras: wire.SnapshotExtras{
			LoginMethod:      &loginMethod,
			AccountEmail:     nil,
			RateLimitTier:    nil,
			ExtraRateWindows: []json.RawMessage{},
		},
		Status: wire.SnapshotStatus{
			State:      wire.StateOK,
			DataSource: wire.DataSourceAPIOnly,
			Stale:      false,
		},
	}, true
}

func rollingWindow(usedPct float64, resetAt time.Time, now time.Time) wire.RollingWindow {
	var resets *string
	if !resetAt.IsZero() {
		s := wire.FormatTime(resetAt)
		resets = &s
	}
	return wire.RollingWindow{
		UsedPct:          usedPct,
		RemainingSeconds: remainingSeconds(resetAt, now),
		ResetsAt:         resets,
	}
}

func quotaWindow(label string, usedPct float64, resetAt time.Time) wire.QuotaWindow {
	var resets *string
	if !resetAt.IsZero() {
		s := wire.FormatTime(resetAt)
		resets = &s
	}
	return wire.QuotaWindow{
		Label:    label,
		Scope:    label,
		UsedPct:  usedPct,
		ResetsAt: resets,
	}
}

func isAggregateCreditLimit(limit upstreamLimit) bool {
	return strings.EqualFold(limit.Source, "aggregate") && strings.EqualFold(limit.LimitType, "credits")
}

func normalizedWindow(window string) string {
	w := strings.ToLower(strings.TrimSpace(window))
	switch w {
	case "5hr", "5hrs", "5hour", "5hours", "primary":
		return "5h"
	case "7day", "7days", "week", "weekly", "secondary":
		return "7d"
	case "":
		return "unknown"
	default:
		return w
	}
}

func parseResetAt(value *string) time.Time {
	if value == nil {
		return time.Time{}
	}
	raw := strings.TrimSpace(*value)
	if raw == "" {
		return time.Time{}
	}
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339} {
		if t, err := time.Parse(layout, raw); err == nil {
			return t
		}
	}
	if seconds, err := strconv.ParseInt(raw, 10, 64); err == nil && seconds > 0 {
		return time.Unix(seconds, 0).UTC()
	}
	return time.Time{}
}

func remainingSeconds(resetAt time.Time, now time.Time) int {
	if resetAt.IsZero() || !resetAt.After(now) {
		return 0
	}
	return int(math.Round(resetAt.Sub(now).Seconds()))
}

func clamp(value, minValue, maxValue float64) float64 {
	if value < minValue {
		return minValue
	}
	if value > maxValue {
		return maxValue
	}
	return value
}

func normalizeBaseURL(raw string) string {
	base := strings.TrimSpace(raw)
	if base == "" {
		return defaultURL
	}
	if parsed, err := url.Parse(base); err == nil {
		parsed.Path = strings.TrimSuffix(parsed.Path, "/v1")
		parsed.RawQuery = ""
		parsed.Fragment = ""
		return strings.TrimRight(parsed.String(), "/")
	}
	return strings.TrimRight(strings.TrimSuffix(base, "/v1"), "/")
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if strings.TrimSpace(value) != "" {
			return value
		}
	}
	return ""
}

File: Sources/TokenUsageCore/State/Snapshot.swift
import Foundation

/// Represents a single quota window in normalized form.
public struct QuotaWindow: Codable, Equatable, Sendable {
    public let label: String
    public let scope: String
    public let usedPct: Double
    public let resetsAt: String?

    enum CodingKeys: String, CodingKey {
        case label
        case scope
        case usedPct = "used_pct"
        case resetsAt = "resets_at"
    }

    public init(label: String, scope: String, usedPct: Double, resetsAt: String?) {
        self.label = label
        self.scope = scope
        self.usedPct = usedPct
        self.resetsAt = resetsAt
    }
}

/// Represents a normalized rolling quota window.
public struct RollingWindow: Codable, Equatable, Sendable {
    public let usedPct: Double
    public let remainingSeconds: Int
    public let resetsAt: String?

    enum CodingKeys: String, CodingKey {
        case usedPct = "used_pct"
        case remainingSeconds = "remaining_seconds"
        case resetsAt = "resets_at"
    }

    public init(usedPct: Double, remainingSeconds: Int, resetsAt: String?) {
        self.usedPct = usedPct
        self.remainingSeconds = remainingSeconds
        self.resetsAt = resetsAt
    }

    public static let empty = RollingWindow(usedPct: 0, remainingSeconds: 0, resetsAt: nil)
}

/// Represents credit balance information when a provider exposes it.
public struct Credits: Codable, Equatable, Sendable {
    public let remaining: Double
    public let updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case remaining
        case updatedAt = "updated_at"
    }

    public init(remaining: Double, updatedAt: String?) {
        self.remaining = remaining
        self.updatedAt = updatedAt
    }
}

/// Carries provider-specific metadata that is not part of the common quota windows.
public struct SnapshotExtras: Codable, Equatable, Sendable {
    public let loginMethod: String?
    public let accountEmail: String?
    public let rateLimitTier: String?
    public let extraRateWindows: [JSONValue]

    enum CodingKeys: String, CodingKey {
        case loginMethod = "login_method"
        case accountEmail = "account_email"
        case rateLimitTier = "rate_limit_tier"
        case extraRateWindows = "extra_rate_windows"
    }

    public init(
        loginMethod: String?,
        accountEmail: String?,
        rateLimitTier: String?,
        extraRateWindows: [JSONValue])
    {
        self.loginMethod = loginMethod
        self.accountEmail = accountEmail
        self.rateLimitTier = rateLimitTier
        self.extraRateWindows = extraRateWindows
    }

    public static let empty = SnapshotExtras(
        loginMethod: nil,
        accountEmail: nil,
        rateLimitTier: nil,
        extraRateWindows: [])
}

/// One quota window for a single claude-swap account.
public struct AccountWindow: Codable, Equatable, Sendable {
    public let usedPct: Double
    public let resetsAt: String?

    enum CodingKeys: String, CodingKey {
        case usedPct = "used_pct"
        case resetsAt = "resets_at"
    }

    public init(usedPct: Double, resetsAt: String?) {
        self.usedPct = usedPct
        self.resetsAt = resetsAt
    }
}

/// One claude-swap-managed Claude account's usage (menu-bar per-account rows).
public struct AccountUsage: Codable, Equatable, Sendable {
    public let number: Int
    public let email: String
    public let active: Bool
    public let status: String
    public let fiveHour: AccountWindow?
    public let sevenDay: AccountWindow?

    enum CodingKeys: String, CodingKey {
        case number
        case email
        case active
        case status
        case fiveHour = "five_hour"
        case sevenDay = "seven_day"
    }

    public init(number: Int, email: String, active: Bool, status: String,
                fiveHour: AccountWindow?, sevenDay: AccountWindow?) {
        self.number = number
        self.email = email
        self.active = active
        self.status = status
        self.fiveHour = fiveHour
        self.sevenDay = sevenDay
    }
}

/// Captures the fetch status embedded in every snapshot.
public struct SnapshotStatus: Codable, Equatable, Sendable {
    public let state: ProviderState
    public let dataSource: SnapshotDataSource
    public let stale: Bool

    enum CodingKeys: String, CodingKey {
        case state
        case dataSource = "data_source"
        case stale
    }

    public init(state: ProviderState, dataSource: SnapshotDataSource = .apiOnly, stale: Bool) {
        self.state = state
        self.dataSource = dataSource
        self.stale = stale
    }
}

/// Represents the final SSE snapshot schema.
public struct UsageSnapshot: Codable, Equatable, Sendable {
    public let schema: Int
    public let seq: Int
    public let generatedAtUTC: String
    public let producerID: String
    public let producerTimeZone: String
    public let provider: Provider
    public let burnRatePerMinute: Double
    public let burnState: String
    public let todayTotalTokens: Int
    public let todaySessions: Int
    public let rolling5h: RollingWindow
    public let weekly: RollingWindow
    public let quotaWindows: [QuotaWindow]
    public let credits: Credits?
    public let extras: SnapshotExtras
    public let status: SnapshotStatus
    public let accounts: [AccountUsage]?
    public let accountsUpdatedAt: String?

    enum CodingKeys: String, CodingKey {
        case schema
        case seq
        case generatedAtUTC = "generated_at_utc"
        case producerID = "producer_id"
        case producerTimeZone = "producer_tz"
        case provider
        case burnRatePerMinute = "burn_rate_per_min"
        case burnState = "burn_state"
        case todayTotalTokens = "today_total_tokens"
        case todaySessions = "today_sessions"
        case rolling5h = "rolling_5h"
        case weekly
        case quotaWindows = "quota_windows"
        case credits
        case extras
        case status
        case accounts
        case accountsUpdatedAt = "accounts_updated_at"
    }

    public init(
        schema: Int = 1,
        seq: Int,
        generatedAtUTC: String,
        producerID: String,
        producerTimeZone: String,
        provider: Provider,
        burnRatePerMinute: Double = 0,
        burnState: String = "idle",
        todayTotalTokens: Int = 0,
        todaySessions: Int = 0,
        rolling5h: RollingWindow,
        weekly: RollingWindow,
        quotaWindows: [QuotaWindow],
        credits: Credits?,
        extras: SnapshotExtras,
        status: SnapshotStatus,
        accounts: [AccountUsage]? = nil,
        accountsUpdatedAt: String? = nil)
    {
        self.schema = schema
        self.seq = seq
        self.generatedAtUTC = generatedAtUTC
        self.producerID = producerID
        self.producerTimeZone = producerTimeZone
        self.provider = provider
        self.burnRatePerMinute = burnRatePerMinute
        self.burnState = burnState
        self.todayTotalTokens = todayTotalTokens
        self.todaySessions = todaySessions
        self.rolling5h = rolling5h
        self.weekly = weekly
        self.quotaWindows = quotaWindows
        self.credits = credits
        self.extras = extras
        self.status = status
        self.accounts = accounts
        self.accountsUpdatedAt = accountsUpdatedAt
    }

    /// Builds a degraded snapshot that preserves the final schema.
    public static func degraded(
        provider: Provider,
        seq: Int,
        producer: ProducerInfo,
        now: Date,
        state: ProviderState)
        -> UsageSnapshot
    {
        UsageSnapshot(
            seq: seq,
            generatedAtUTC: SnapshotDateFormatter.string(from: now),
            producerID: producer.id,
            producerTimeZone: producer.timeZone,
            provider: provider,
            rolling5h: .empty,
            weekly: .empty,
            quotaWindows: [],
            credits: nil,
            extras: .empty,
            status: SnapshotStatus(state: state, stale: true))
    }
}
